Get rid of the const field in CTransaction

sipa opened the pull request:
As explained on https://github.com/zcash/zcash/issues/967, using const member fields in CTransaction (or any deserializable class) is not well defined in C++, as such fields should only ever be assigned at construction time.

I believe we can get rid of them. Their purpose was making sure that CTransaction was not used in places where mutable fields were required, but now that CMutableTransaction is well-established that requirement is not so strong anymore.

Longer term, I'd prefer to see better encapsulation of CTransaction (which could also introduce better internal representation of the data, without a gazillion allocations) and improvements to the serialization framework so that we can construct objects by deserializing... but I think that's further out than a fix for this.

So at a higher level, this changes CTransaction from "immutable transaction data" to "transaction data with cached hashes". I've added a small comment about this as well.

Ping @daira @ebfull.

sipa committed:
Get rid of the const field in CTransaction

ebfull commented:
utACK

dcousens commented:
utACK 232529f

laanwj commented:
From a design point I really liked CTransaction as immutable transaction data, but yes it's better to be without the C++ hacks.

dcousens commented:
Would this also negate the need for `CMutableTransaction`? (which is only used in `bitcoin-tx`)

sipa commented:
@dcousens It's used in all signing code. And CMutableTransaction has the advantage of not needing to call UpdateHash all the time, so it may be useful to keep.

sipa commented:
@laanwj I agree, it gave a pretty clean and safe abstraction... but I know no way to do it correctly with the current serialization framework. We should think about a construct-while-deserializing mechanism if we want that.

daira commented:
NACK due to https://github.com/sipa/bitcoin/commit/232529fbf1af2261085e45407611c32f08d5f6f3#commitcomment-18556288 (easily fixed).

sipa commented:
@daira In my understanding (but I may be wrong), a reference to a const value only means that that value cannot be modified _through_ that reference but does not mean that it cannot change overall. This is different from declaring a variable or field as const, which means it cannot change at all.

sipa commented:
@daira I addressed all your comments apart from the returning of const uint256&.

This issue made me investigate the concept of const correctness in C++ much more deeply, but all explanations I can find state that a const reference merely means that the value aliased by that reference cannot be modifed _through that reference_. It is not a guarantee that the value cannot be modified at al. Consider the example at the bottom of http://en.cppreference.com/w/cpp/language/const_cast; a const reference is taken of a non-const value; then it is modified (even through the const reference, which is legal as the caller knows a non-const alias (`i`) exists as well), and accessed through the const reference. If your reasoning was correct, this example would be undefined.

daira commented:
Yes, you're right. (Also see https://isocpp.org/wiki/faq/const-correctness#aliasing-and-const .) utACK.

sipa commented:
See #8580 for an alternative solution that is a better step towards making CTransaction having its own more efficient in-memory representation.

laanwj commented:
Closing in favor of #8580
